### PR TITLE
Implement str.rindex() based on CPython

### PR DESCRIPTION
**Pull request by densmirn**

### CODE REVIEW

**Comment by densmirn** [comment]
Could you please queue the PR for review?

File: numba/unicode.py [hunk 1]
```diff
@@ -509,6 +509,26 @@ def find_impl(a, b):
         return find_impl
 
 
+def unicode_idx_check_type(ty, name):
```
Comment (esc) [inline numba/unicode.py]: this seems to be a duplication of the `unicode_rfind_check_type` below. Can you re-use this function there too and reduce duplication.

Comment (densmirn) [reply numba/unicode.py]: Reused that in https://github.com/numba/numba/pull/4866.

**Comment by densmirn** [comment]
After updating the PR from master CI fails. Could you please fix it?

**Comment by esc** [comment]
@densmirn we just saw it too, investigating now.

**Comment by densmirn** [comment]
> @densmirn we just saw it too, investigating now.

Thanks!

**Comment by stuartarchibald** [comment]
@densmirn you don't have permission to restart Azure directly, but as a workaround, just closing and reopening the PR will re-trigger CI. Current master has a bug on it, it's being fixed right now :)

**Comment by densmirn** [comment]
Could you please continue reviewing the PR?

**Comment by esc** [comment]
So yes, either we merge this PR in the next merge window and you submit a new PR that  unifies the tests for `index` and `rindex` or you can make the changes to the test suite here now.

**Comment by densmirn** [comment]
> So yes, either we merge this PR in the next merge window and you submit a new PR that unifies the tests for `index` and `rindex` or you can make the changes to the test suite here now.

Let me unify tests for `index` and `rindex` in separate PR and merge this PR as is.

**Comment by stuartarchibald** [comment]
Please could you resolve the conflicts by merging master into this branch. Thanks!

**Comment by densmirn** [comment]
Resolved all the merge conflicts.

**Pull request merged**